### PR TITLE
feature: BlankLineBetweenImportGroupsFixer - Introduction

### LINKED ISSUE #4821: Empty line between import groups
Hello,

Redorder of the `use` statements in the header block seems not to take into account the "use function ..." and it order it like other use.

After using `@PhpCsFixer` set, it end up with:
```
use FATBundle\Service;
use function GuzzleHttp\json_encode;
use Pimcore\Controller\FrontendController;
```

Should be:
```
use FATBundle\Service;
use Pimcore\Controller\FrontendController;
use function GuzzleHttp\json_encode;
```

It would be great for lisibility !

---

If we go further, I think it should be spaced, for better visibility:
```
use FATBundle\Service;
use Pimcore\Controller\FrontendController;

use function GuzzleHttp\json_encode;
```

_I know you are working on PSR12 rules, so it would allow PSR12 compliance (i run this tool to check: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/PSR12/Sniffs/Files/FileHeaderSniff.php#L357 , aka rule `Header blocks must be separated by a single blank line`). Please see https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements. But PSR12 is another subject, but as this issue is about talking use statement I mention it._

I think `use function...` (and `use const...`) should be ordered at end and space by default.

Best regards.

### PR DESCRIPTION
**Pull request by SanderVerkuil**
Closes #4821 

I've also updated the PSR12 set to include this rule, to be compliant with the [PSR12](https://www.php-fig.org/psr/psr-12/#3-declare-statements-namespace-and-import-statements).

### CODE REVIEW

File: tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+{
+    /**
+     * @dataProvider provideFixTypesOrderAndWhitespaceCases
+     */
+    public function testFixTypesOrderAndNewlines(string $expected, ?string $input = null): void
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public function provideFixTypesOrderAndWhitespaceCases(): array
```
Comment (GrahamCampbell) [inline tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php:34]: I think we need some tests with some funky comments too, to make sure we don't accidentality move some code inside a comment.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php (hunk 1 above)
Comment (SanderVerkuil) [reply tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php:34]: I have added some test cases, and I hope they're funky enough with the comments. Though I am not entirely sure what the actual expected behavior ought to be, as the PSR12 doesn't allow comments within the use blocks if I recall correctly.

So if you have any more test cases, please do let me know. I hope the current tests are sufficient.

**Review by julienfalque — APPROVED** [review]
Awesome! Just a few nitpicks and this can be merged IMO.

File: src/Fixer/Whitespace/BlankLineBetweenImportGroupsFixer.php [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+use PhpCsFixer\Tokenizer\Tokens;
+use PhpCsFixer\Tokenizer\TokensAnalyzer;
+
+/**
+ * @author Sander Verkuil <s.verkuil@pm.me>
+ */
+final class BlankLineBetweenImportGroupsFixer extends AbstractFixer implements WhitespacesAwareFixerInterface
+{
+    /**
+     * @internal
```
[the remark this answers is not in the document]
Comment (SanderVerkuil) [reply src/Fixer/Whitespace/BlankLineBetweenImportGroupsFixer.php:33]: Alright, has been changed 💪

File: src/Fixer/Whitespace/BlankLineBetweenImportGroupsFixer.php [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
+        $namespacesImports = $tokensAnalyzer->getImportUseIndexes(true);
+
+        if (0 === \count($namespacesImports)) {
```
[the remark this answers is not in the document]
Comment (SanderVerkuil) [reply src/Fixer/Whitespace/BlankLineBetweenImportGroupsFixer.php]: This block has been removed, thanks for the feedback!

**Review by julienfalque — APPROVED** [review]

**Comment by SanderVerkuil** [comment]
@SpacePossum and/or @keradus, sorry for the ping. This is my first pull request to contribute to the amazing PHP CS Fixer, and I was wondering what further needs to be done to get this approved and being used upstream?

File: tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php (hunk 1 above)
Comment (jrmajor) [reply tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php:34]: I don't think you have any tests with comments inside `use` statements:

```php
use /* x */ function /* x */ Psl\Str\trim;
use /* x */ Psl\Str /* x */;
use /* x */ const /* x */ Psl\Str\ALPHABET;
use /* x */ function /* x */ Psl\Str\ /* x */ {
    /* x */ trim_left /* x */,
    /* x */ trim_right /* x */,
};
```

Comment (jrmajor) [reply tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php:34]: Wow, I've just discovered that you can even do this (it imports `Psl\Str\trim_left`):

```php
use function P/**/s/**/l/**/\/**/S/**/t/**/r/**/\/**/t/**/r/**/i/**/m/**/_/**/l/**/e/**/f/**/t/**/;
```

Probably no need to test that in this fixer though :)

Comment (SanderVerkuil) [reply tests/Fixer/Whitespace/BlankLineBetweenImportGroupsFixerTest.php:34]: I've added the tests for the initial tests with the comments inside the use statement. In PHP 7, the following block results in a syntax error, so I couldn't perform these.

```php
use function P/**/s/**/l/**/\/**/S/**/t/**/r/**/\/**/t/**/r/**/i/**/m/**/_/**/l/**/e/**/f/**/t/**/;
```

**Comment by SpacePossum** [comment]
The PR already looks great, I left a little review PR on your fork to deal some edge cases where the fixer would break the code, let me know what you think.

**Review by SpacePossum — APPROVED** [review]

**Comment by julienfalque** [comment]
Hi @SanderVerkuil. Some tests are failing, can you please fix them? Then we can merge :)

**Comment by SanderVerkuil** [comment]
> Hi @SanderVerkuil. Some tests are failing, can you please fix them? Then we can merge :)

Hmn, @SpacePossum added some tests where comments have been added within the use import statement, like:

```php
use Weird\Be/**comment*/hat;
```

Which is apparently correct php in some cases, but not all.

The errors are regarding PHP syntax errors, and used to run fine on PHP 7 if I recall correctly, but apparently break on PHP 8.1. Should I just remove those test cases (as the error is that it's a PHP syntax error), or only test these cases in PHP 7?

**Comment by julienfalque** [comment]
Not sure this test has much value if it is only valid on a PHP version that we'll likely stop supporting soon anyway. I would remove it. What do you think @SpacePossum?

**Comment by SpacePossum** [comment]
The failure on 7.4 are a priority test/fixture issue not related to the case:
```php
use Weird\Be/**comment*/hat;
```

So lets see what is up with that.

Than lets move the `use Weird\Be/**comment*/hat;` to a dedicated test 
```
    /**
     * @requires PHP <8.0
     */
```
, as I don't know what the planning is for dropping 7.4 support but my guess is it might be for a while

**Comment by SpacePossum** [comment]
https://github.com/SanderVerkuil/PHP-CS-Fixer/pull/2

**Comment by SanderVerkuil** [comment]
@SpacePossum Will do that as soon as I'm home tomorrow 💪. Thanks for your help and support! Now only another approval for the tests and I'll happily squash all commits

**Comment by SanderVerkuil** [comment]
@SpacePossum The commits have been rebased and squashed, and is now a single commit. Thanks for all the help, let's hope this will be merged soon 🤞

**Comment by SpacePossum** [comment]
Thanks! : D

**Comment by julienfalque** [comment]
Thank you @SanderVerkuil.
